internal/layers/dataset.go

```diff
@@ -0,0 +1,236 @@
+package layers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/pgxpool"
+	"github.com/mimiro-io/internal-go-util/pkg/uda"
+	"github.com/mimiro-io/postgresql-datalayer/internal/conf"
+	"github.com/mimiro-io/postgresql-datalayer/internal/db"
+	"strings"
+)
+
+const batchChunkSize = 1000
+
+type ReadableDataset interface {
+	Read(ctx context.Context, entities chan<- *uda.Entity) error
+	ReadChanges(ctx context.Context, since string, entities chan<- *uda.Entity) (string, error)
+}
+
+type WriteableDataset interface {
+	Write(ctx context.Context, entities <-chan *uda.Entity) error
+}
+
+type PostgresDataset struct {
+	pg         *pgxpool.Pool
+	table      *db.ReadTable
+	writeTable *db.WriteTable
+	since      string
+	limit      int64
+}
+
+func NewPostgresDataset(pg *pgxpool.Pool, table *db.ReadTable, writeTable *db.WriteTable, request db.DatasetRequest) *PostgresDataset {
+	return &PostgresDataset{
+		pg:         pg,
+		table:      table,
+		writeTable: writeTable,
+		since:      request.Since,
+		limit:      request.Limit,
+	}
+}
+
+// Write takes a chan of uda.Entity and queues this in a batch request. Once the batch request has
+// batchChunkSize in entities, it attempts to commit the batch in a transaction. If the transaction
+// fails, an error is returned from the writer, and control is returned to the caller.
+func (ds *PostgresDataset) Write(ctx context.Context, entities <-chan *uda.Entity) error {
+	defer func() {
+		ds.pg.Close() // since write creates a new pool, we should close it when done
+	}()
+	if ds.writeTable == nil { // we could do a table check against the database here
+		return errors.New("missing write table")
+	}
+	// we want to chunk this into blocks of data, then commit each chunk
+	chunk := make([]*uda.Entity, batchChunkSize)
```

Comment:
I wonder if we should try to keep the transaction alive for the whole duration of the post-request. sending batches is still a good thing. but maybe not create a new transaction for each batch/chunk? 

I realize that it is not likely to cause data consistency issues to commit partial payloads. the committed rows will just be overwritten with the same data one more time when the post-request is repeated later. Merely wonder if it would be a cleaner approach to either write all or rollback all.

Reply:
Not disagreeing with you per se, I'm just a bit worried about the potential in memory consequences of holding several millions of rows of insert queries in a slice. So I elected to keep it like it is....

Reply:
yeah there are problems like transaction timeouts as well.